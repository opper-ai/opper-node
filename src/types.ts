export type Options = {
    /**
     * OpperAI [API key](https://docs.opper.ai/api-keys)
     */
    apiKey: string;

    /**
     * By default, client-side use of this library is not allowed, as it risks exposing your secret API credentials to attackers.
     * Only set this option to `true` if you understand the risks and have appropriate mitigations in place.
     */
    dangerouslyAllowBrowser?: boolean;

    /**
     * Override the base url if needed
     * Defaults to https://api.opper.ai
     */
    baseURL?: string;

    isUsingAuthorization?: boolean;
};

export type Message = {
    role: "user" | "assistant";
    content: string;
};

export interface Chat {
    parent_span_uuid?: string;
    path: string;
    message: string | Message[];
    examples?: OpperExample[];
}

export interface SSEStreamCallbacks {
    controller?: AbortController | null;
    onMessage: (json: unknown) => void;
    onComplete: () => void;
    onError: (error: Error) => void;
    onCancel?: () => void;
}

export interface OpperAIStream {
    path: string;
    parent_span_uuid?: string;
    message: string | Message[];
    callbacks: SSEStreamCallbacks;
}

export interface OpperChatResponse {
    span_id: string;
    message: string;
    json_payload?: Record<string, unknown>;
    context: unknown;
    error?: string;
    cached?: boolean;
}

export interface OpperImageResponse {
    bytes: Uint8Array;
}

export type OpperIndexFile = {
    uuid: string;
    original_filename: string;
    size: number;
    index_status:
        | "init"
        | "pending"
        | "uploading"
        | "indexing"
        | "success"
        | "indexed"
        | "error"
        | "failed"
        | "invalid";
    n_vectors: number;
};

export type OpperIndex = {
    uuid: string;
    name: string;
    created_at: Date;
    files: OpperIndexFile[];
};

export type OpperIndexFilterOperation = "=" | "!=" | ">" | "<" | "in";

export interface OpperIndexFilter {
    key: string;
    operation: OpperIndexFilterOperation;
    value: string | number | Array<string | number>;
}

export type OpperIndexDocument = {
    uuid?: string;
    key?: string; // a unique key that one can use if you want to update the document
    content: string;
    metadata: Record<string, unknown>;
};

export type OpperIndexQuery = {
    /**
     * The query to retrieve documents for.
     */
    query: string;
    /**
     * The maximum number of documents to retrieve.
     */
    k: number;
    /**
     * The filters to apply to the documents in the index.
     */
    filters?: OpperIndexFilter[];
    /**
     * The parent span uuid to use for the request.
     */
    parent_span_uuid?: string;
};

export type OpperExample = {
    input: unknown;
    output: unknown;
    comment?: string;
};

export type OpperGenerateImage = {
    prompt: string;
    image_size?: string;
    quality?: string;
    parent_span_uuid?: string;
};

export type OpperCall = {
    /**
     * The name of the function to be used in the Opper UI.
     */
    name: string;

    /**
     * The input to the function.
     */
    input: string | Record<string, unknown>;

    /**
     * The instructions for the call sent to the model to be used as part of the prompt.
     * Defaults to "you are a helpful assistant".
     */
    instructions?: string;

    /**
     * The model to use to generate the output.
     * See: https://docs.opper.ai/functions/models
     */
    model?: string;

    /**
     * The input schema for the function.
     */
    input_schema?: Record<string, unknown>;

    /**
     * The output schema for the function.
     */
    output_schema?: Record<string, unknown>;

    /**
     * This is used to attach to an existing span.
     * If not provided, a new span will be created and displayed in the Opper UI.
     */
    parent_span_uuid?: string;

    /**
     * Manual examples to use as part of the prompt to guide the model.
     */
    examples?: OpperExample[];

    /**
     * Whether to stream the response from the function.
     */
    stream?: boolean;

    /**
     * Configuration for the function.
     */
    configuration?: {
        invocation?: {
            /**
             * Use few shot prompting (also know as In Context Learning) to enable Opper to automatically pull in
             * semantically relevant examples that will guide the model.
             * See: https://docs.opper.ai/datasets
             */
            few_shot?: {
                /**
                 * The number of few shot examples to use.
                 */
                count?: number;
            };
        };
    };
};

export type OpperFunction = {
    uuid?: string;
    path: string;
    description: string;
    instructions: string;
    model?: string;
    index_ids?: number[];
    few_shot?: boolean;
    few_shot_count?: number;
    cache_config?: CacheConfig;
    input_schema?: Record<string, unknown>;
    out_schema?: Record<string, unknown>;
    dataset_uuid?: string;
};

export interface CacheConfig {
    exact_match_cache_tll: number;
    semantic_cache_threshold: number;
    semantic_cache_tll: number;
}

export type Span = {
    uuid: string;
    name?: string;
    input?: string;
    output?: string;
    start_time?: Date;
    parent_uuid?: string;
    end_time?: Date;
    error?: string;
    meta?: Record<string, unknown>;
    evaluations?: Record<string, unknown>;
    score?: number;
};

export type SpanMetric = {
    /**
     * Dimension of the metric. E.g. "accuracy", "faithfulness", "fluency"
     */
    dimension?: string;
    /**
     * Score of the metric. E.g. 0.95
     */
    score?: number;
    /**
     * Comment of the metric. E.g. "The answer is correct"
     */
    comment?: string;
};

export type Generation = {
    called_at?: Date;
    duration_ms: number;
    model?: string;
    response?: string;
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
    error?: string;

    messages?: Message[];
    cost?: number;
};

export type Dataset = {
    uuid: string;
    name: string;
    description?: string;
    project_uuid: string;
    created_at?: Date;
    updated_at?: Date;
};

export type DatasetEntry = {
    uuid: string;
    input: string;
    output: string;
    comment?: string;
    span_id?: string;
    trace_id?: string;
    created_at?: Date;
    updated_at?: Date;
};
