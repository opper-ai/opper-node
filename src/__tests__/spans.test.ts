import Client from "../index";
import Spans from "../spans";
import { SpanMetric } from "../types";

// Mocking the global fetch to avoid actual API calls
// @ts-expect-error Mocking global fetch
global.fetch = jest.fn(() =>
    Promise.resolve({
        json: () => Promise.resolve({}),
    })
);

describe("Spans", () => {
    let traces: Spans;
    const mockApiKey = "test-api-key";

    beforeEach(() => {
        const mockClient = new Client({
            apiKey: mockApiKey,
        });

        jest.useFakeTimers();
        jest.setSystemTime(new Date(2020, 3, 1));

        traces = new Spans(mockClient);
        // Clear all instances and calls to constructor and all methods:
        (global.fetch as jest.Mock).mockClear();
    });

    describe("startSpan", () => {
        it("should call fetch with correct parameters and resolve with the current span", async () => {
            const mockSpan = {
                uuid: "span-uuid",
                name: "test-span",
                start_time: new Date(),
            };
            (global.fetch as jest.Mock).mockResolvedValueOnce({
                ok: true,
                status: 200,
                json: () => Promise.resolve({ uuid: mockSpan.uuid }),
            });

            const spanId = await traces.startSpan(mockSpan);

            expect(global.fetch).toHaveBeenCalledWith(expect.stringContaining("/v1/spans"), {
                method: "POST",
                headers: expect.any(Object),
                body: expect.any(String),
            });
            expect(spanId).toMatchSnapshot();
        });
    });

    describe("endSpan", () => {
        it("should call fetch with correct parameters to update the span", async () => {
            const mockSpan = {
                uuid: "span-uuid",
                name: "test-span",
                start_time: new Date(),
                end_time: new Date(),
            };
            (global.fetch as jest.Mock).mockResolvedValueOnce({
                ok: true,
                status: 200,
                json: () => Promise.resolve({ uuid: mockSpan.uuid }),
            });

            const spanId = await traces.endSpan(mockSpan);

            expect(global.fetch).toHaveBeenCalledWith(
                expect.stringContaining(`/v1/spans/${mockSpan.uuid}`),
                {
                    method: "PUT",
                    headers: expect.any(Object),
                    body: expect.any(String),
                }
            );
            expect(spanId).toMatchSnapshot();
        });
    });

    describe("saveMetric", () => {
        it("should call fetch with correct parameters and resolve with metric UUID", async () => {
            const spanUuid = "span-uuid";
            const mockFeedback: SpanMetric = {
                dimension: "quality",
                score: 0.9,
                comment: "Good job",
            };
            (global.fetch as jest.Mock).mockResolvedValueOnce({
                ok: true,
                status: 200,
                json: () => Promise.resolve({ uuid: "metric-uuid" }),
            });

            const metricId = await traces.saveMetric(spanUuid, mockFeedback);

            expect(global.fetch).toHaveBeenCalledWith(
                expect.stringContaining(`/v1/spans/${spanUuid}/metrics`),
                {
                    method: "POST",
                    headers: expect.any(Object),
                    body: expect.any(String),
                }
            );
            expect(metricId).toEqual("metric-uuid");
        });
    });

    describe("saveExample", () => {
        it("should call fetch with correct parameters and resolve with example UUID", async () => {
            const spanUuid = "span-uuid";
            (global.fetch as jest.Mock).mockResolvedValueOnce({
                ok: true,
                status: 200,
                json: () => Promise.resolve({ uuid: "example-uuid" }),
            });

            const exampleId = await traces.saveExample(spanUuid);

            expect(global.fetch).toHaveBeenCalledWith(
                expect.stringContaining(`/v1/spans/${spanUuid}/save_examples`),
                {
                    method: "POST",
                    headers: expect.any(Object),
                    body: undefined,
                }
            );
            expect(exampleId).toEqual("example-uuid");
        });
    });

    describe("span with child span", () => {
        it("Should create a parent and child span", async () => {
            (global.fetch as jest.Mock).mockResolvedValue({
                ok: true,
                status: 200,
                json: () => Promise.resolve({ uuid: "return-a-uuid" }),
            });

            const parent = await traces.startSpan({
                name: "parent-span",
                input: "parent input",
                uuid: "parent-uuid",
            });

            const child = await traces.startSpan({
                name: "child-span",
                uuid: "child-uuid",
                input: "child input",
                parent_uuid: "parent-uuid",
            });

            const childEndSpan = await traces.endSpan({ ...child, output: "child output" });
            const parentEndSpan = await traces.endSpan({ ...parent, output: "parent output" });

            expect(childEndSpan.name).toBe("child-span");
            expect(childEndSpan.output).toBe("child output");
            expect(childEndSpan.parent_uuid).toBe("parent-uuid");
            expect(childEndSpan.project).toBe("missing_project");
            expect(childEndSpan).toMatchSnapshot();

            expect(parentEndSpan.name).toBe("parent-span");
            expect(parentEndSpan.output).toBe("parent output");
            expect(parentEndSpan.parent_uuid).toBe(undefined);
            expect(parentEndSpan.project).toBe("missing_project");
            expect(parentEndSpan).toMatchSnapshot();
        });
    });
});
