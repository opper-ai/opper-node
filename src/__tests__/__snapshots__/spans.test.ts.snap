// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spans endSpan should call fetch with correct parameters to update the span 1`] = `
{
  "end_time": 2020-04-01T00:00:00.000Z,
  "name": "test-span",
  "parent_uuid": undefined,
  "start_time": 2020-04-01T00:00:00.000Z,
  "uuid": "span-uuid",
}
`;

exports[`Spans span with child span Should create a parent and child span 1`] = `
{
  "end_time": 2020-04-01T00:00:00.000Z,
  "input": "child input",
  "name": "child-span",
  "output": "child output",
  "parent_uuid": "parent-uuid",
  "project": "missing_project",
  "start_time": 2020-04-01T00:00:00.000Z,
  "uuid": "return-a-uuid",
}
`;

exports[`Spans span with child span Should create a parent and child span 2`] = `
{
  "end_time": 2020-04-01T00:00:00.000Z,
  "input": "parent input",
  "name": "parent-span",
  "output": "parent output",
  "parent_uuid": undefined,
  "project": "missing_project",
  "start_time": 2020-04-01T00:00:00.000Z,
  "uuid": "return-a-uuid",
}
`;

exports[`Spans startSpan should call fetch with correct parameters and resolve with the current span 1`] = `
{
  "input": "",
  "name": "test-span",
  "project": "missing_project",
  "start_time": 2020-04-01T00:00:00.000Z,
  "uuid": "span-uuid",
}
`;
